File: core/conn/jdbcT4/src/main/java/org/trafodion/jdbc/t4/SecPwd.java
/**********************************************************************
// @@@ START COPYRIGHT @@@
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
// @@@ END COPYRIGHT @@@
 //
 **********************************************************************/

/**
 * class SecPwd - builds the password key, encrypts password,
 *                creates HMAC message based on password, rolename
 *                process info and time stamp.  It also gets expiration
 *                date of a certificate.
 *
 */

package org.trafodion.jdbc.t4;

import java.io.File;
import java.io.FileOutputStream;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;
import java.sql.Connection;


public class SecPwd {
	 /**
	  *
      *
      * @return SecPwd
      */
     public static SecPwd getInstance(Connection con, String directory, String fileName,
    		 String serverName, boolean spjMode, byte[] procInfo) throws SecurityException
     {
    	 if (con == null)
    		 throw new SecurityException(SecClientMsgKeys.INPUT_PARAMETER_IS_NULL, new Object[]{"connection"});
    	 SecPwd secpwd = new SecPwd(directory, fileName, serverName, spjMode, procInfo);

    		 SymCrypto.insert(con, secpwd);


    	 return secpwd;
     }

     public static void removeInstance(Connection con)
     {
    	 SymCrypto.remove(con);
     }

	/**
	 * Ctor for the SecPwd. There are two possible certificates: active
	 * certificate and certificate that is going to be active.
	 *
	 * If autodownload is true, certificate will always come from the
	 * server. In this case, only active certificate is used.
	 *
	 * If autodownload is false, active certificate is used to encrypt the
	 * password. When there is a new certificate, it will be stored in
	 * "certificate". As soon as this new certificate is activated on the
	 * server, the current active certificate will become stale, and the new
	 * certificate will be copied over and becomes the active certificate.
	 *
	 * If spjMode is true, the OS name is NONSTOP_KERNEL and the host name
	 * is the same as the server name then just setSpj mode to true
	 * and does nothing.
	 *
	 * @param directory
	 *            specifies the directory to locate the certificate. The default
	 *            value is %HOME% if set else %HOMEDRIVE%%HOMEPATH%.
	 * @param fileName
	 *            specifies the certificate that is in waiting. The default
	 *            value is the first 5 characters of server name.
	 * @param activeFileName
	 *            specifies the current certificate in use. The default value is
	 *            the first 5 character of server name + Active
	 * @param spjMode
	 *            true - and if os.name == NSK and the host name
	 *            matches the local host - token case.  Certificate is not
	 *            handled in this case.
	 *            false - handles certificate
	 * @param serverName
	 *            server name for this certificate.
	 * @throws SecurityException
	 */
	private SecPwd(String directory, String fileName, 
			String serverName, boolean spjMode, byte[] procInfo) throws SecurityException {
		// check USERID env variable for MXCI testing of SPJs.  If set use normal password
		// encryption
		if ((spjMode == true)  &&
			//	((hostName.substring(0, 5)).compareToIgnoreCase(serverName.substring(0, 5)) == 0) &&
				(System.getenv("USERID") == null))// token
		{
			m_spjMode = spjMode;
		}
		else // password
		{
			if (procInfo == null)
				throw new SecurityException(SecClientMsgKeys.INPUT_PARAMETER_IS_NULL, new Object[]{"procInfo"});
			// Stores procInfo with the time stamp for data message encryption used
			m_procInfo = new byte [SecdefsCommon.PROCINFO_SIZE + SecdefsCommon.TIMESTAMP_SIZE];
			System.arraycopy(procInfo, 0, m_procInfo, 0, (SecdefsCommon.PROCINFO_SIZE + SecdefsCommon.TIMESTAMP_SIZE));
			directory = (directory != null) ? directory : System.getenv("HOME");
			if (directory == null)
			{
				String hmdrive = System.getenv("HOMEDRIVE");
				String hmpath = System.getenv("HOMEPATH");
				if (hmdrive != null && hmpath != null)
				{
					directory = hmdrive + File.separator + hmpath;
				}
			    else
			    {
			    	directory = System.getProperty("user.home");
			    	if (directory == null)
			    		throw new SecurityException (SecClientMsgKeys.HOME_ENVIRONMENT_VAR_IS_NULL, null);
			    }
            }
			fileName = (fileName != null) ? fileName : serverName + ".cer";

			File dir = new File(directory);
			if (dir.isDirectory() == false)
				throw new SecurityException(SecClientMsgKeys.DIR_NOTFOUND, new Object[]{dir.getPath()});

			certFile = new File(directory, fileName);
		}
	}

	/**
	 * Processes the active certificate when spjMode is false
	 * else does nothing.  The certificate is processed by calling
	 * the Security ctor to creates the password key and initializes it
     * with password id.  Gets public key and the length of the public
     * key from the certificate file.  Generates nonce and session key.
	 * @throws SecurityException
	 */
	public void openCertificate() throws SecurityException {
		if (m_spjMode == false) // do nothing for the token case
			m_sec = new Security(certFile);
	}

	/** This method builds the password key which consists 4 bytes of password id,
	 *  128 bytes of role name which would be 128 spaces when role name is null,
	 *  32 bytes of the digest message calculated using the session key on the data made up of
	 *  the procInfo and the encrypted data and 256 bytes (if the 2048 public key is used) or
	 *  128 bytes (if the1024 public key is used) encrypted data calculated using the public key
	 *  on the plain text made up of the session key, the nonce and the password.
	 *  The password key is generated only when the spjMode is false.  When
	 *  the spjMode is true, 26 bytes of the token is returned instead.
	 * Builds password key
	 * @param pwd
	 * 		 	password to be encrypted
	 * @param rolename
	 * 			role name to build password key
	 * @param procInfo
	 * 			process information (PIN, CPU, segment name and time stamp)
	 * @return pwdkey
	 * 			returns the password key if spjMode is false
	 *          returns the token when spjMode is true
	 * @throws SecurityException
	 */

	public void encryptPwd(byte[] pwd, byte[] rolename, byte[] pwdkey) throws SecurityException {
		// rolename is optional so can be NULL
		if (pwd == null)
			throw new SecurityException(SecClientMsgKeys.INPUT_PARAMETER_IS_NULL, new Object[]{"password"});
		if (pwdkey == null)
			throw new SecurityException(SecClientMsgKeys.INPUT_PARAMETER_IS_NULL, new Object[]{"password key"});
		if (m_spjMode == true) // token
		{
			if (pwd.length != SecdefsCommon.TOKENSIZE)
				throw new SecurityException(SecClientMsgKeys.BAD_TOKEN_LEN, null);
			if ((pwd[0] != SecdefsCommon.USERTOKEN_ID_1)
					|| (pwd[1] != SecdefsCommon.USERTOKEN_ID_2))
				throw new SecurityException(
						SecClientMsgKeys.INCORRECT_TOKEN_FORMAT, null);
			ByteBuffer.wrap(pwd).get(pwdkey, 0, SecdefsCommon.TOKENSIZE);
		}
		else
		{
			m_sec.encryptPwd(pwd, rolename, m_procInfo, pwdkey);
		}
	}

	/** Gets length of buffer for password encryption (public)
	 * or the length of the token if it is the SPJ mode
	 * @returns
	 *     If the spjMode is false
	 *        the length of the password key is returnd if success
	 * 	      0 if failed
	 *     If spjMode is true
	 *     	  the length of the token is returned
	 * @throws SecurityException
	 */
	public int getPwdEBufferLen() throws SecurityException {
		if (m_spjMode == true)
			return SecdefsCommon.TOKENSIZE;
		else
			return m_sec.getPwdEBufferLen();
	}

	/** Gets the expiration date of the certificate
	 * @return an array of bytes
	 * 			presents the certificate's
	 * 			expiration day in the format YYMMDDHHMMSS
	 * 			or a zero length byte array if the it is in the SPJ mode
	 */
	public byte[] getCertExpDate() {
		if (m_spjMode == false)
			return m_sec.getCertExpDate();
		else
			return new byte[0];
	}

	/**
	 * When autodownload is on, client will download the certificate from server
	 * when there is no certificate or certificate is stale.
	 *
	 * @param buf
	 *            content of the certificate pushed from server.
	 */
	public void switchCertificate(byte[] buf) throws SecurityException {
		FileChannel outChannel = null;
		try {
			outChannel = new FileOutputStream(certFile).getChannel();
			outChannel.write(ByteBuffer.wrap(buf));
		} catch (Exception e) {
			throw new SecurityException(e, SecClientMsgKeys.ERR_WRITE_CERT_FILE, new Object[]{certFile});
		} finally {
			try {
				if (outChannel != null)
					outChannel.close();
			} catch (Exception e) {
			}
		}
		m_sec = new Security(certFile);
	}

	public byte[] getProcInfo()
	{
		return m_procInfo;
	}

	private Security m_sec;
	private File certFile;
	private boolean m_spjMode;
	private byte[] m_procInfo;   //stores only 4 bytes pid + 4 bytes nid


};

File: win-odbc64/security_dll/native/source/secpwd.cpp
/*************************************************************************
// @@@ START COPYRIGHT @@@
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
// @@@ END COPYRIGHT @@@
 **************************************************************************/
//
#include <stdio.h>
#include <string.h>
#include <sys/stat.h>
#include "security.h"
#include "secpwd.h"
#include "utils.h"
#include "drvrglobal.h"


#ifdef _WINDOWS
#define SEPARATOR "\\"
#else
#define SEPARATOR "/"
#endif

const int BUFSIZE = 1024;
    static void copyFile(char* source, char* target)
throw (SecurityException)
{
    FILE *in = fopen(source, "r");
    if ( in == NULL )
        throw SecurityException(ERR_READ_CERT_FILE, source);

    FILE *out = fopen(target, "w");
    if ( out == NULL )
        throw SecurityException(ERR_WRITE_CERT_FILE, target);

    char buf[BUFSIZE];
    while ( !feof(in) )
    {
        int bread = fread(buf, 1, BUFSIZE, in);
        if (bread == 0) break;
        int bwrite = fwrite(buf, 1, bread, out);
        if ( bread != bwrite )
        {
            fclose(in);
            fclose(out);
            throw SecurityException(ERR_WRITE_CERT_FILE, target);
        }
    }
    fclose(in);
    fclose(out);
}

static int compareFile(char* source, char* target)
{
    FILE *in = fopen(source, "r");
    FILE *out = fopen(target, "r");

    if ( in == NULL || out == NULL ) return 1;

    int retVal = 0;
    char inContent[BUFSIZE], outContent[BUFSIZE];
    while ( !feof(in) )
    {
        int bread_in = fread(inContent, 1, BUFSIZE, in);
        int bread_out = fread(outContent, 1, BUFSIZE, out);

        if ( bread_in != bread_out ||
                memcmp(inContent, outContent, bread_in) )
        {
            retVal = 1;
            break;
        }
    }
    fclose(in);
    fclose(out);
    return retVal;
}

static void TrimRight(char* str)
{
    // this function trims any unwanted spaces at the end of a string.

    char *buffer = str;

    int len = (int)strlen(buffer);
    int index = 0;

    for (index = len - 1; index >= 0; index--) {
        if ((index == 0) && (buffer[index] == ' ')) {
            buffer[index] = '\0';
            break;
        }
        if (buffer[index] != ' ') {
            buffer[index+1] = '\0';
            break;
        }
    }
}

static void TrimLeft(char *str)
{

    // this function trims any unwanted spaces at the beginning of a string.
    int len = strlen(str);
    char *buffer = new char[len + 1];
    strcpy(buffer, str);

    int index = 0;

    for (index = 0; index < len; index++) {
        if ((index == len - 1) && (buffer[index] == ' ')) {
            buffer[0] = '\0';
            strcpy(str ,buffer);
            break;
        }
        if (buffer[index]!= ' ') {
            strcpy(str,buffer+index);
            break;
        }
    }

    delete buffer;
}

#define CER ".cer"
#define ACTIVE_CER ".cer"

static char* buildName(const char* dir, const char* fileName, const char* serverName, const char* suffix)
{
    char *fname = NULL;
    int srvrNmLen = strlen(serverName);
    int sfxLen = strlen(suffix);
    int dirLen = strlen(dir);

    int len = dirLen + strlen(SEPARATOR);
    int fNmLen = srvrNmLen + sfxLen + 1;

    fname = new char[fNmLen];   // Runtime determine the buffer length of the cert file name

    if ( fileName == NULL )
    {
        strcpy_s(fname, fNmLen, serverName);
        strcat_s(fname, fNmLen, suffix);
        fileName = fname;
        len += strlen(fname) + 1;
    }
    else len += strlen(fileName) + 1;

    char* path = new char[len];
    sprintf(path, "%s%s%s\0", dir, SEPARATOR, fileName);
    delete[] fname;

    return path;
}

SecPwd::SecPwd(const char *dir, const char* fileName,
        const char* activeFileName,
        const char* serverName,
        int serverNameLength) throw (SecurityException)
: m_sec(NULL), certFile(NULL), activeCertFile(NULL)
{
    char* dir_from_env = NULL;
    char  certDir[MAX_SQL_IDENTIFIER_LEN + 1];
    char* serverNameGBKToUtf8 = NULL;
    LCID lcid = GetSystemDefaultLCID();
    int  translen = 0;
    char transError[128] = { 0 };

    if (serverName == NULL)
        throw SecurityException (INPUT_PARAMETER_IS_NULL, " - serverName.");

    dir = (dir != NULL) ? dir : getenv("HOME");

    if (dir == NULL)
    {
        //Check HOMEDRIVE (or HOMEDRIVE + HOMEPATH ) for interix
        char *hmdrive=getenv("HOMEDRIVE");
        char *hmpath=getenv("HOMEPATH");
        if (hmdrive != NULL && hmpath != NULL)
        {
            // Fix solution #10-090803-8661
            TrimLeft(hmdrive);
            TrimRight(hmdrive);
            TrimLeft(hmpath);
            TrimRight(hmpath);
            dir_from_env = new char[strlen(hmdrive) + strlen(hmpath) + 1];
            sprintf((char *)dir_from_env, "%s%s\0", hmdrive, hmpath);
        }
        else
            throw SecurityException (HOME_ENVIRONMENT_VAR_IS_NULL, NULL);
    }
    else
    {
        TrimLeft((char *)dir);
        TrimRight((char *)dir);
    }
    if(dir != NULL)
        strncpy(certDir, dir, sizeof(certDir));
    else
    {
        strncpy(certDir, dir_from_env, sizeof(certDir));
        certDir[sizeof(certDir) -1] = '\0';
        delete[] dir_from_env;
    }

    struct stat st;
    if(stat(certDir,&st) != 0)
        throw SecurityException(DIR_NOTFOUND, (char *)certDir);

    if (lcid == 0x804) // if local charset is gbk
    {
        serverNameGBKToUtf8 = (char *)malloc(MAX_SQL_IDENTIFIER_LEN + 1);
        if (TranslateUTF8(TRUE, serverName, serverNameLength,
            serverNameGBKToUtf8, MAX_SQL_IDENTIFIER_LEN, &translen, transError) != SQL_SUCCESS)
        {
            free(serverNameGBKToUtf8);
            serverNameGBKToUtf8 = NULL;
            throw SecurityException(INPUT_PARAMETER_IS_NULL, " - serverName.");
        }
        certFile = buildName(certDir, fileName, serverNameGBKToUtf8, CER);
        activeCertFile = buildName(certDir, activeFileName, serverNameGBKToUtf8, ACTIVE_CER);

        free(serverNameGBKToUtf8);
        serverNameGBKToUtf8 = NULL;
    }
    else
    {
        certFile = buildName(certDir, fileName, serverName, CER);
        activeCertFile = buildName(certDir, activeFileName, serverName, ACTIVE_CER);
    }

    // If activeCertFile does not exist and certFile exist, create activeCertFile
    // by from copying certFile to activeCertFile
    struct stat certStats;
    struct stat aCertStats;

    if ( (stat(certFile, &certStats) == 0 ) &&
            (stat(activeCertFile, &aCertStats) != 0 ))
        copyFile(certFile, activeCertFile);
}

void SecPwd::openCertificate() throw (SecurityException)
{
    if(m_sec != NULL)
    {
        delete m_sec;
        m_sec = NULL;
    }
    m_sec = new Security(activeCertFile);
}

SecPwd::~SecPwd()
{
    delete certFile;
    delete activeCertFile;
    if(m_sec != NULL)
    {
        delete m_sec;
        m_sec = NULL;
    }
}
void SecPwd::encryptPwd(const char *pwd, const int pwdLen,
        const char *rolename, const int rolenameLen,
        const char *procInfo, const int procInfoLen,
        char *pwdkey, int *pwdkeyLen)
throw (SecurityException)
{
    // rolename is optional so can be NULL
    if (!pwd)
        throw SecurityException(INPUT_PARAMETER_IS_NULL, " - pwd.");
    if (!pwdkey)
        throw SecurityException(INPUT_PARAMETER_IS_NULL, " - pwdkey.");
    if (!procInfo)
        throw SecurityException(INPUT_PARAMETER_IS_NULL, " - procInfo.");
    if (*pwdkeyLen != m_sec->getPwdEBufferLen())
        throw SecurityException(BAD_PWDKEY_LENGTH, NULL);

    // If Little Endian, swap the 2 cpu bytes and the 2 pin bytes of the PprocInfo
    if (!m_sec->getLittleEndian())
    {
        ProcInfo *pI = (ProcInfo *) procInfo;
        ProcInfo *pInfo = new ProcInfo;

        pInfo->cpu = swapByteOrderOfS((unsigned int)pI->cpu);
        pInfo->pin = swapByteOrderOfS((unsigned int)pI->pin);
        //	  memcpy((char*)pInfo->seg_name, (char *)pI->seg_name, sizeof(pInfo->seg_name));
        pInfo->timestamp = (long long)swapByteOrderOfLL((unsigned long long)pI->timestamp);
        m_sec->encryptPwd(pwd, pwdLen, rolename, rolenameLen,
                (char *)pInfo, procInfoLen, pwdkey, pwdkeyLen);
        delete pInfo;
    }
    else
        m_sec->encryptPwd(pwd, pwdLen, rolename, rolenameLen,
                procInfo, procInfoLen, pwdkey, pwdkeyLen);
}


int SecPwd::getPwdEBufferLen() throw (SecurityException)
{
    return (m_sec->getPwdEBufferLen());
}

unsigned char* SecPwd::getCertExpDate()
{
    return m_sec->getCertExpDate();
}


    void SecPwd::switchCertificate(char* content, int len)
throw (SecurityException)
{
    FILE *out = fopen(activeCertFile, "w");
    if ( out == NULL )
        throw SecurityException(ERR_WRITE_CERT_FILE, activeCertFile);

    int bwrite = fwrite(content, 1, len, out);

    fclose(out);
    if ( bwrite != len )
        throw SecurityException(ERR_WRITE_CERT_FILE, activeCertFile);

    if(m_sec != NULL)
    {
        delete m_sec;
        m_sec = NULL;
    }
    m_sec = new Security(activeCertFile);
}

int SecPwd::switchCertificate() throw (SecurityException)
{
    // If same name, then no point trying to switch.
    if (strcmp(certFile, activeCertFile) == 0) return 1;

    struct stat inFileStats;
    struct stat outFileStats;
    int retVal = 0;

    if ( stat(certFile, &inFileStats) == 0 )
    {
        if ( stat(activeCertFile, &outFileStats) == 0 )
        {
            // if both files exist and they differ, then copy
            if ( inFileStats.st_size != outFileStats.st_size ||
                    compareFile(activeCertFile, certFile)   )
            {
                copyFile(certFile, activeCertFile);
            }
            else
            {
                // file with same content, don't switch.
                retVal = 1;
            }
        }
        else
        {
            // activeCertFile doesn't exist, copy from certFile
            copyFile(certFile, activeCertFile);
        }
    }
    else retVal = 1;

    if ( retVal == 0 )
    {
        delete m_sec;
        m_sec = new Security(activeCertFile);
    }
    return retVal;
}

char* SecPwd::getCertFile()
{
    return certFile;
}

char* SecPwd::getActiveCertFile()
{
    return activeCertFile;
}

